### DISCUSSION TITLE
How do I get the local_key ?

### DISCUSSION THREAD

**Discussion by ratnick** [jasonacox/tinytuya#111] [Q&A] [answered]
I'm struggling to get the local_key for my smart plug.
I have tried using the tinytuya wizard as suggested, but I just get 
"Error from Tuya server: your ip(80.1xx.xxx.113) don't have access to this API". I am using my WAN IP address (which I have whitelisted on the Tuya cloud).

I have seen a few link, suggesting to download other apps to get it, but I guess there must be an easy way. Or?

Thanks.

**Comment by jasonacox** [comment]
Hi @ratnick - this means there is some issue with the configuration on the Tuya IoT Cloud setup.  The following things tend to cause this:

## API Subscribed and Authorized

It could be that the API is not subscribed or authorized - Ensure you have subscribed and authorized to these API service:

<img width="583" alt="image" src="https://user-images.githubusercontent.com/836718/156294522-0ae7bb79-250a-4fa4-afa8-aec45dffe4d8.png">

How to: 
SERVICE API: Under "Service API" ensure these APIs are listed: IoT Core, Authorization and Smart Home Scene Linkage. To be sure, click subscribe again on every service. Very important: disable popup blockers otherwise subscribing won't work without providing any indication of a failure. Make sure you authorize your Project to use those APIs:
Click "Service API" tab
Click "Go to Authorize" button
Select the API Groups from the dropdown and click Subscribe ([screenshot](https://user-images.githubusercontent.com/38729644/128742724-9ed42673-7765-4e21-94c8-76022de8937a.png))

## Datacenter

If your IP does not match the datacenter region you selected you will not be able to access the API.  The data center regions are not logical, unfortunately.  You may need to try multiple datacenters until you get one to work.  Some have reported success in selecting all the datacenters in the Tuya Cloud setup and then trying different datacenters in the Wizard.

I hope this helps.

**Reply by ratnick** [reply]
Thanks for answering. My problem turned out to be something else: I confused the different terms. For future reference, I provide my setup below.

**Comment by ratnick** [comment, marked as answer]
I did the following to make it work. 

Based on: https://github.com/jasonacox/tinytuya

# d = tinytuya.OutletDevice('DEVICE_ID_HERE', 'IP_ADDRESS_HERE', 'LOCAL_KEY_HERE')
DEVICE_ID: found throug Smart Life app or through the cloud.
IP_ADDRESS: This is the LOCAL IP address, for example found found via IP Scanner. The Smart Life App gives you the global address - don't use that.
LOCAL_KEY: Same as "key" found down below through the tinytuya wizard (in Python).

But this wizard needs some other input (apiKey and apiSecret), which I found via the project set up in the Tuya cloud (as described in https://github.com/jasonacox/tinytuya):

**Cloud -> Development -> select the project you have created.** 
**On this page, apiKey = Access ID, and apiSecret=Access Secret**.  (yes - one wculd wish for consistent terminology)
And by the way: If you, as I was, are in doubt about which data server you should be on, look here: https://developer.tuya.com/en/docs/iot/oem-app-data-center-distributed?id=Kafi0ku9l07qb

Next, start with a CLI. Note: If you see crappy text (escape characters etc.), your CLI cannot handle colors. Download "ansicon" instead.

Go to the place where you installed ansicon (the json files will be stored here by default. You can also move the ansicon files. 	
>>> python -m tinytuya wizard
(first time you run, enter the keys from above. Second time, just edit the tinytuya.json file in the same directory, listed here:
Edit tinytuya.json
		{
		    "apiKey": "xxxxxyyyyyowaafwa7g0",
		    "apiSecret": "xxxxxyyyyy5c4ee5bd2749442178a646",
		    "apiRegion": "eu",
		    "apiDeviceID": "xxxxxyyyyy0aaf15a2uxd3"
		}
	API Key=xxxxxyyyyyowaafwa7g0 (Access ID, covers all APIs attached to the Tuya Cloud project)
	Secret=xxxxxyyyyy5c4ee5bd2749442178a646 (Access Secret)
	DeviceID=xxxxxyyyyy0aaf15a2uxd3 (via Smart Things app or via "Devices" in the cloud project)

The output from the wizard:
python -m tinytuya wizard
		TinyTuya Setup Wizard [1.3.1]
		
		    Existing settings:
		        API Key=xxxxxxxxxxxwaafwa7g0
		        Secret=xxxxxxxxxxxxxx5bd2749442178a646
		        DeviceID=xxxxxxxxxxxxxxxc32jwku
		        Region=eu
		
		    Use existing credentials (Y/n):
		Device Listing
		[
		    {
		        "name": "Wifi Plug 3",
		        "id": "xxxxxxxxxxxx8663acc32jwku",
		        "key": "xxxxxxc61bb977383"      ####### Here's your LOCAL_KEY to use in the top
		    },
		    {
		        "name": "havepumpe",
		        "id": "xxxxxxxxxx7c01c3361mx6q",
		        "key": "xxxxxxxxxxe92e2e199e"
		    },
		    {
		        "name": "party lamper",
		        "id": "ccxxxxxxxxxxx04c3e3f8bamww",
		        "key": "xxxxxxxxxx287388b"
		    }
		]
		
		>> Saving list to devices.json
		    3 registered devices saved
		
		>> Saving raw TuyaPlatform response to tuya-raw.json

**Reply by jasonacox** [reply]
> Note: If you see crappy text (escape characters etc.), your CLI cannot handle colors. Download "ansicon" instead.

For Windows 10+ systems you can also run this to make ANSI work in command prompt: 
```ps
reg add HKEY_CURRENT_USER\Console /v VirtualTerminalLevel /t REG_DWORD /d 0x00000001 /f
```

**Reply by jasonacox** [reply]
Thanks for the steps @ratnick !
